extension/popup/popup.jsx

```diff
@@ -0,0 +1,744 @@
+/* globals util, voice, vad, lottie, settings, log, voiceShim, buildSettings */
+
+const React = window.React
+const { Component, PureComponent } = React
+const ReactDOM = window.ReactDOM
+
+const PERMISSION_REQUEST_TIME = 2000
+const FAST_PERMISSION_CLOSE = 500
+let stream
+let isWaitingForPermission = null
+let executedIntent = false
+
+const { backgroundTabRecorder } = buildSettings
+
+// Default amount of time (in milliseconds) before the action is automatically dismissed after we perform certain actions (e.g. successfully switching to a different open tab). This value should give users enough time to read the content on the popup before it closes.
+const DEFAULT_TIMEOUT = 2500
+// Timeout for the popup when there's text displaying:
+const TEXT_TIMEOUT = 7000
+let overrideTimeout = null
+
+let recorder
+let recorderIntervalId
+
+const initialState = {
+    currentState: "listening",
+    suggestions: [],
+    transcript: null,
+    error: null,
+    displayText: null,
+    displayAutoplay: false,
+    search: null
+}
+
+class Popup extends Component {
+    constructor(props) {
+        super(props)
+        this.state = {...initialState}
+        this.textInputDetected = false
+    }
+
+    componentDidMount() {
+        document.addEventListener("keydown", this.onKeyPressed)
+        this.init()
+    }
+
+    componentWillUnmount() {
+        document.removeEventListener("keydown", this.onKeyPressed)
+
+        browser.runtime.sendMessage({ type: "microphoneStopped" })
+        if (!executedIntent) {
+            browser.runtime.sendMessage({ type: "cancelledIntent" })
+        }
+        if (
+            !backgroundTabRecorder &&
+            isWaitingForPermission &&
+            Date.now() - isWaitingForPermission < FAST_PERMISSION_CLOSE
+        ) {
+            this.startOnboarding()
+        }
+
+        // TODO: offload mic and other resources before closing?
```

Comment:
Does this all get run even when the window is closed? I don't think so, and so I think most of this cleanup should happen in a `beforeunload` handler. (Probably this component will never unmount?)